src/ucp/wireup/wireup.c

```diff
@@ -1351,28 +1352,77 @@ static void ucp_wireup_discard_uct_eps(ucp_ep_h ep, uct_ep_h *uct_eps,
     }
 }
 
+/* Check if AM lane should be flushed after being operational (wireup process
+ * completed) */
+static int
+ucp_wireup_is_am_need_flush(ucp_ep_h ep, const ucp_lane_index_t *reuse_lane_map)
+{
+    ucp_lane_index_t am_lane = ucp_ep_get_am_lane(ep);
+
+    /* If CM lane exists, no AM messages were sent through AM lane, thus flush
+     * is not required */
+    return !ucp_ep_has_cm_lane(ep) &&
+           /* Verify AM lane exists */
+           (am_lane != UCP_NULL_LANE) &&
+           /* Lane is not being reused */
+           (reuse_lane_map[am_lane] == UCP_NULL_LANE) &&
+           /* Lane is not yet operational */
+           !ucp_ep_is_lane_p2p(ep, am_lane);
+}
+
 static int
 ucp_wireup_check_is_reconfigurable(ucp_ep_h ep,
                                    const ucp_ep_config_key_t *new_key,
                                    const ucp_unpacked_address_t *remote_address,
                                    const unsigned *addr_indices)
 {
-    ucp_lane_index_t lane;
+    ucp_lane_index_t lane, wireup_lane, reuse_lane_map[UCP_MAX_LANES];
+    const ucp_ep_config_key_t *old_key;
 
     if ((ep->cfg_index == UCP_WORKER_CFG_INDEX_NULL) ||
         ucp_ep_has_cm_lane(ep)) {
         return 1;
     }
 
-    /* TODO: Support reconfiguration when lanes are created without a wireup_ep
-     * wrapper */
-    for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
-        if (!ucp_wireup_ep_test(ucp_ep_get_lane(ep, lane))) {
+    old_key = &ucp_ep_config(ep)->key;
+
+    /* TODO: 1) Support lanes which are connected to the same remote MD, but
+     *          different remote sys_dev (eg. TCP). */
+    for (lane = 0; lane < old_key->num_lanes; ++lane) {
+        if ((ucp_ep_config_find_match_lane(old_key, lane, new_key) !=
+             UCP_NULL_LANE) &&
+            !ucp_ep_config_lane_is_equal(old_key, new_key, lane)) {
```

Comment:
`!ucp_ep_config_find_match_lane` seems redundant?

Reply:
`ucp_ep_config_find_match_lane` detects equal dst_md_index and `ucp_ep_config_lane_is_equal` detects different sys_dev

Reply:
comparing of `dst_md_index` is a part of `ucp_ep_config_lane_is_equal` as well as `dst_sys_dev`

Reply:
when `ucp_ep_config_lane_is_equal` returns false, it is still unknown if md1 and md2 are equal.
If they are not equal, `ucp_wireup_check_is_reconfigurable` should return 1

Reply:
Ok, just too many NOT in the sentence and conditions... maybe this could improve readability?
```
    for (lane = 0; lane < old_key->num_lanes; ++lane) {
        if (ucp_ep_config_lane_is_equal(old_key, new_key, lane) || // <-- check this first for perf reason
            (ucp_ep_config_find_match_lane(old_key, lane, new_key) == UCP_NULL_LANE)) { 
            continue;
        }

        return 0;
    }

```